src/plans/D3MAaveTypeBufferPlan.sol

```diff
@@ -0,0 +1,111 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import "./ID3MPlan.sol";
+
+interface TokenLike {
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface ATokenLike {
+    function UNDERLYING_ASSET_ADDRESS() external view returns (address);
+    function ATOKEN_REVISION() external view returns (uint256);
+}
+
+/**
+ *  @title D3M Aave Buffer Plan
+ *  @notice Ensure `buffer` amount of DAI is always available to borrow.
+ *  @dev This plan can be used with both V2 and V3 versions of the Aave codebase.
+ */
+contract D3MAaveTypeBufferPlan is ID3MPlan {
+
+    mapping (address => uint256) public wards;
+    uint256                      public buffer;
+
+    TokenLike  public immutable dai;
+    ATokenLike public immutable adai;
+    uint256    public immutable adaiRevision;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, uint256 data);
+
+    constructor(address adai_) {
+        adai = ATokenLike(adai_);
+        dai = TokenLike(adai.UNDERLYING_ASSET_ADDRESS());
+        adaiRevision = adai.ATOKEN_REVISION();
+        
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MAaveTypeBufferPlan/not-authorized");
+        _;
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "buffer") {
+            buffer = data;
+        } else revert("D3MAaveTypeBufferPlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function getTargetAssets(uint256 currentAssets) external override view returns (uint256) {
+        if (buffer == 0) return 0; // De-activated
+
+        uint256 liquidityAvailable = dai.balanceOf(address(adai));
+        if (buffer >= liquidityAvailable) {
+            // Need to increase liquidity
+            return currentAssets + (buffer - liquidityAvailable);
+        } else {
+            // Need to decrease liquidity
+            unchecked {
+                uint256 decrease = liquidityAvailable - buffer;
+                if (currentAssets >= decrease) {
+                    return currentAssets - decrease;
+                } else {
+                    return 0;
+                }
+            }
+        }
+    }
+
+    function active() public view override returns (bool) {
+        if (buffer == 0) return false;
+        uint256 adaiRevision_ = adai.ATOKEN_REVISION();
+        return adaiRevision_ == adaiRevision;
```

Comment:
```suggestion
        return buffer > 0 && adai.ATOKEN_REVISION() == adaiRevision;
```

Reply:
Btw did we consider removing the adai revision and make this plan more standard?
I guess that was your first idea as you had mentioned it wouldn't be specific for any project, but then something changed that this was considered necessary.

Reply:
Yeah I'm not sure if it adds much. Lots of things can change and really all it needs to do is look like a rebasing ERC-20. Any change that would break the assumptions here would be the same as if Aave put through a malicious change or something (IE breaking everything).

Reply:
Yeah for the purpose of the plan, we just need a "source" address where to check the actual liquidity available we don't care what that token is. And if we want to check some malicious thing, then that adai_revision check is quite poor. So not sure, I would tend to remove it and do something standard for any project.

Reply:
@iamchrissmith @rockyfour opinions on this one before thinking on any change?

Reply:
Yeah I guess I agree with the reasoning and it can be removed.

Reply:
Ok will remove.

Reply:
Done https://github.com/makerdao/dss-direct-deposit/pull/102/commits/991d96935aa90ddf25b7f8e7e70cabb17c45701b

Reply:
Now it is removed, we could actually just make this a standard plan, where we remove the `adai` reference and we pass in the constructor `dai` and a `gem` (`adai` for this implementation).
This way the plan would be just `D3MBufferPlan`

Reply:
It is still pretty specific to Aave in that they store their measure of liquidity inside the aToken.

Reply:
Why not call it `src`. So basically what you need is the source where to check liquidity. You do not care if it is a token or what. In this case will be `adai` but in other could be a pool address.

Reply:
I'm saying calling `dai.balanceOf(src)` is still assuming the dai balance == available liquidity. It's possible there is additional accounting.

Reply:
I mean, my main idea to remove that validation was to make this general. But if you feel strongly about keeping as aavetype I wouldn't fight for the opposite either.

Reply:
I think you are right, wouldn't work even for Compound, as the totalPoolSize is defined as the balance of dai in cdai + borrows/debt, but subtracts the totalReserves. So that should also be affecting the available liquidity. Let's leave it as it is.